abdiff/core/calc_ab_metrics.py

```diff
@@ -0,0 +1,293 @@
+    )
+
+    return metrics_data
+
+
+def create_record_diff_matrix_parquet(
+    run_directory: str,
+    diffs_dataset: str,
+    batch_size: int = 1_000,
+) -> str:
+    """Create a boolean sparse matrix of modified fields for all records.
+
+    This writes a single parquet file with rows for each record, and columns for each
+    TIMDEX field, and a value of integer 1 if that field has a diff and 0 if not.  This
+    provides a handy way to calculate aggregate metrics for a given field or source in
+    later steps.
+
+    This code does NOT use pyarrow, and momentarily creates a single dataframe in memory
+    for all rows.  This is safe given the nature of the dataframe: there may be 10m rows,
+    but there are only 20-30 columns, and either integer 1 or 0 as the value, resulting
+    in a very small dataset in memory despite a potentially high row count.
+    """
+    diffs_ds = ds.dataset(diffs_dataset)
+
+    batch_metrics_dfs = []
+    for i, batch in enumerate(
+        diffs_ds.to_batches(
+            batch_size=batch_size,
+            columns=["timdex_record_id", "source", "ab_diff"],
+        )
+    ):
+        t0 = time.time()
+        batch_df = batch.to_pandas()
+
+        # parse diff JSON to dictionary for batch
+        batch_df["ab_diff"] = batch_df["ab_diff"].apply(
+            lambda diff_json: json.loads(diff_json)
+        )
+
+        batch_metrics = []
+        for _, row in batch_df.iterrows():
+            record_metrics = {
+                "timdex_record_id": row["timdex_record_id"],
+                "source": row["source"],
+            }
+            diff_data = row["ab_diff"]
+            record_metrics.update(get_record_field_diff_bools_for_record(diff_data))
+            batch_metrics.append(record_metrics)
+
+        # build dataframe for batch
+        batch_metrics_df = pd.DataFrame(batch_metrics)
+        batch_metrics_dfs.append(batch_metrics_df)
+        logger.info(f"batch: {i+1}, elapsed: {time.time()-t0}")
+
+    # concatenate all dataframes into single dataframe for writing and replace None with 0
+    metrics_df = pd.concat(batch_metrics_dfs)
+    metrics_df = metrics_df.fillna(0)
+
+    # write parquet file via pandas writer
+    filepath = Path(run_directory) / "metrics" / "metrics.parquet"
+    metrics_df.to_parquet(filepath, index=False)
+    return str(filepath)
+
+
+def get_record_field_diff_bools_for_record(diff_data: dict) -> dict:
+    """Function to return dictionary of fields that have a diff.
+
+    Determining if a field had a diff is as straight-forward as looking to see if it shows
+    up in the parsed diff JSON.  The fields may be at the root of the diff, or they could
+    be nested under "$insert" or "$delete" nodes in the diff.
+
+    If a field from the original A/B records are not in the diff at all, then they did not
+    have changes, and therefore will not receive a 1 here to indicate a diff.
+    """
+    fields_with_diffs = {}
+    seen_subfields: set[str] = set()
```

Comment:
Hmm, how is this variable used? I don't see any values being added to the set in the `for` loop. 🤔

Reply:
Nice catch!  This was some carryover from a first approach, where it's no longer needed.  Removing.